funsor/pyro/convert.py

```diff
@@ -65,28 +67,47 @@ def funsor_to_tensor(funsor_, ndims, event_inputs=()):
     return tensor
 
 
+@singledispatch
 def dist_to_funsor(pyro_dist, event_inputs=()):
     """
     Convert a :class:`torch.distributions.Distribution` to a
     :class:`~funsor.terms.Funsor` .
     """
     assert isinstance(pyro_dist, torch.distributions.Distribution)
+    raise ValueError("Cannot convert {} distribution to a Funsor"
+                     .format(type(pyro_dist).__name__))
 
-    while isinstance(pyro_dist, dist.Independent):
-        pyro_dist = pyro_dist.base_dist
 
-    if isinstance(pyro_dist, dist.Categorical):
-        return tensor_to_funsor(pyro_dist.logits, event_inputs + ("value",))
+@dist_to_funsor.register(dist.Independent)
+def _independent_to_funsor(pyro_dist, event_inputs=()):
+    event_names = tuple("_event_{}".format(len(event_inputs) + i)
+                        for i in range(pyro_dist.reinterpreted_batch_ndims))
+    result = dist_to_funsor(pyro_dist.base_dist, event_inputs + event_names)
+    for name in reversed(event_names):
+        result = Independent(result, "value", name)
+    return result
 
-    if isinstance(pyro_dist, dist.Normal):
-        loc = tensor_to_funsor(pyro_dist.loc, event_inputs)
-        scale = tensor_to_funsor(pyro_dist.scale, event_inputs)
-        return Normal(loc, scale)
 
-    if isinstance(pyro_dist, dist.MultivariateNormal):
-        loc = tensor_to_funsor(pyro_dist.loc, event_inputs, 1)
-        scale_tril = tensor_to_funsor(pyro_dist.scale_tril, event_inputs, 2)
-        return MultivariateNormal(loc, scale_tril)
+@dist_to_funsor.register(dist.Categorical)
+def _categorical_to_funsor(pyro_dist, event_inputs=()):
+    return tensor_to_funsor(pyro_dist.logits, event_inputs + ("value",))
```

Comment:
Why not return a `funsor.distributions.Categorical` here, as you did with `dist.Normal` below?  If possible I would like to confine the distribution special-casing to a single point in the conversion pipeline, ideally within `funsor.distributions` (which already handles conversion of `Categorical`s to `Tensor`s and `Normal`s to `Gaussian`s)

Reply:
The reason is that `Categorical` currently requires `probs` rather than `logits`, whereas the rest of our HMM-like code works with logits. Eventually I'd like to fork `Categorical` into `CategoricalProbs`+`CategoricalLogits` as we've done with `Bernoulli`, but that seemed out of scope for this PR.